lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3790,6 +3812,12 @@ static FuncDecl *findAutoDiffDerivativeFunction(
   return funcDecl;
 }
 
+void AttributeChecker::visitDerivativeAttr(DerivativeAttr *attr) {
+  if (typeCheckDerivativeAttr(Ctx, D, attr))
```

Comment:
Are `@derivative` attributes in the primary file already type-checked (e.g. parameter indices are resolved) after `TypeChecker::typeCheckDerivativeAttrs` is called in `TypeCheckSourceFileRequest::evaluate`?

If so, is `AttributeChecker::visitDerivativeAttr` doing redundant work (e.g. recomputing parameter indices)?

If so, can `AttributeChecker::visitDerivativeAttr` be changed to assert that `attr` is already type-checked, and to avoid redundant work by only checking for duplicate `@derivative` attributes?

Reply:
Yes, there is redundant work happening. I can't deduplicate the work by doing exactly what you say because the second typechecking uses some of the intermediate calculations to make diagnostic messages (e.g. original function not found messages). So I need to either redo the computation or find some way of passing the intermediate calculations from the first time it typechecks to the second time it typechecks.

Here's an idea that could work:
```
static bool typeCheckDerivativeAttr(...) {
  if (!attr->getOriginalFunction() || !attr->getParameterIndices()) {
    // Do typechecking
    // If successful, set original function and parameter indices
    // If failure, diagnose and return true.
  }

  // Insert attr into Ctx.DerivativeAttrs
  // Diagnose duplicates
}
```

That way, if the first round of typechecking is successful, then we don't duplicate the work on the second round. If the first round fails, we duplicate the work during the second round, but speed is less important in the failure case so this isn't terrible.

Reply:
> I can't deduplicate the work by doing exactly what you say because the second typechecking uses some of the intermediate calculations to make diagnostic messages (e.g. original function not found messages).

Aha, that makes sense.

> Here's an idea that could work:
> 
> ```
> static bool typeCheckDerivativeAttr(...) {
>   if (!attr->getOriginalFunction() || !attr->getParameterIndices()) {
>     // Do typechecking
>     // If successful, set original function and parameter indices
>     // If failure, diagnose and return true.
>   }
> 
>   // Insert attr into Ctx.DerivativeAttrs
>   // Diagnose duplicates
> }
> ```
> 
> That way, if the first round of typechecking is successful, then we don't duplicate the work on the second round. If the first round fails, we duplicate the work during the second round, but speed is less important in the failure case so this isn't terrible.

This idea sounds great! Duplicate work only for bad attributes seems very acceptable.